File: java/src/test/java/com/aaronicsubstances/shrewd/evolver/CustomLogEventTest.java
package com.aaronicsubstances.shrewd.evolver;

import java.util.List;

import static java.util.Arrays.asList;
import static com.aaronicsubstances.shrewd.evolver.TestUtils.toMap;

import org.testng.annotations.DataProvider;
import org.testng.annotations.Test;

import static org.testng.Assert.*;

public class CustomLogEventTest {
    
    public void testAddProperty() {
        CustomLogEvent instance = new CustomLogEvent();
        assertEquals(instance.getData(), null);

        instance.addProperty("age", 20);
        assertEquals(instance.getData(), toMap("age", 20));

        instance.addProperty("name", "Kofi");
        assertEquals(instance.getData(), toMap("age", 20, "name", "Kofi"));
    }

    public void testGenerateMessage() {
        CustomLogEvent instance = new CustomLogEvent();
        instance.setData(toMap("person", toMap("name", "Kofi"),
            "age", "twenty"));

        instance.generateMessage((j, s) -> j.apply("person/name") + " is " +
            s.apply("age") + " years old.");
        assertEquals(instance.getMessage(), "\"Kofi\" is twenty years old.");
    }
    
    
    @Test(dataProvider = "creatTestGetTreeDataSliceData")
    public void testGetTreeDataSlice(Object treeData, String path, Object expected) {
        CustomLogEvent instance = new CustomLogEvent();
        instance.setData(treeData);
        Object actual = instance.fetchDataSlice(path);
        assertEquals(actual, expected);
    }

    @DataProvider
    public Object[][] creatTestGetTreeDataSliceData() {
        return new Object[][]{
            { null, "", null },
            { "", "", "" },
            { 2, "", 2 },
            { asList(2), "f", null },
            { asList(2), "0", 2 },
            { asList(21, asList()), "10", null },
            { asList(21, asList()), "0", 21 },
            { asList(21, asList()), "1", asList() },
            { asList(21, asList()), "-1", asList() },
            { asList(21, asList()), "-2", null },
            { toMap("a", 1), "", toMap("a", 1) },
            { toMap("a", 1), "0", null },
            { toMap("a", 1), "a", 1 },
            { toMap("a", 1, "b", 2), "b", 2 },
            { toMap("a", 1, "b", 2), "c", null },
            { toMap("a", 1, "b", asList("e", true)), "b/0", "e" },
            { toMap("a", 1, "b", asList("e", true)), "b/1", true },
            { toMap("a", 1, "b", toMap("e", true)), "b", toMap("e", true) },
            { toMap("a", 1, "b", toMap("e", true)), "b/e", true }
        };
    }
}

File: csharp/ShrewdEvolver.UnitTests/CustomLogEventTest.cs
using System;
using System.Collections.Generic;
using System.Dynamic;
using System.Text;
using Xunit;
using static AaronicSubstances.ShrewdEvolver.UnitTests.TestUtils;

namespace AaronicSubstances.ShrewdEvolver.UnitTests
{
    public class CustomLogEventTest
    {
        [Fact]
        public void TestAddProperty()
        {
            var instance = new CustomLogEvent(GetType());
            Assert.Null(instance.Data);

            instance.AddProperty("age", 20);
            Assert.Equal(new Dictionary<string, object>
            {
                { "age", 20 }
            }, instance.Data);

            instance.AddProperty("name", "Kofi");
            Assert.Equal(new Dictionary<string, object>
            {
                { "age", 20 },
                { "name", "Kofi" }
            }, instance.Data);
        }

        [Fact]
        public void TestGenerateMessage()
        {
            var instance = new CustomLogEvent(GetType());
            dynamic properties = new ExpandoObject();
            instance.Data = properties;
            properties.person = new Dictionary<string, string>
            {
                { "name", "Kofi" }
            };
            properties.age = "twenty";

            instance.GenerateMessage((j, s) => $"{j($"person/name")} is " +
                $"{s($"age")} years old.");
            Assert.Equal("\"Kofi\" is twenty years old.", instance.Message);
        }

        [Theory]
        [MemberData(nameof(CreatTestGetTreeDataSliceData))]
        public void TestGetTreeDataSlice(object treeData, string path, object expected)
        {
            var instance = new CustomLogEvent(null)
            {
                Data = treeData
            };
            object actual = instance.FetchDataSlice(path);
            Assert.Equal(expected, actual);
        }

        public static List<object[]> CreatTestGetTreeDataSliceData()
        {
            return new List<object[]>
            {
                new object[]{ null, "", null },
                new object[]{ "", "", "" },
                new object[]{ 2, "", 2 },
                new object[]{ ToList(2), "f", null },
                new object[]{ ToList(2), "0", 2 },
                new object[]{ ToList(21, ToList()), "10", null },
                new object[]{ ToList(21, ToList()), "0", 21 },
                new object[]{ ToList(21, ToList()), "1", ToList() },
                new object[]{ ToList(21, ToList()), "-1", ToList() },
                new object[]{ ToList(21, ToList()), "-2", null },
                new object[]{ ToDict("a", 1), "", ToDict("a", 1) },
                new object[]{ ToDict("a", 1), "0", null },
                new object[]{ ToDict("a", 1), "a", 1 },
                new object[]{ ToDict("a", 1, "b", 2), "b", 2 },
                new object[]{ ToDict("a", 1, "b", 2), "c", null },
                new object[]{ ToDict("a", 1, "b", ToList("e", true)), "b/0", "e" },
                new object[]{ ToDict("a", 1, "b", ToList("e", true)), "b/1", true },
                new object[]{ ToDict("a", 1, "b", ToDict("e", true)), "b", ToDict("e", true) },
                new object[]{ ToDict("a", 1, "b", ToDict("e", true)), "b/e", true }
            };
        }
    }
}
